### PR TITLE
[GCS][Storage unification 3/n] Deprecate MemoryInternalKV, use StoreClientInternalKV<MemoryStoreClient> instead

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

In this PR, we deprecated MemoryInternalKV. Instead, we use a wrapper that wraps around MemoryStoreClient. This PR is build upon #23754 and #23725

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 1]
```diff
@@ -446,7 +447,8 @@ void GcsServer::InitKVManager() {
   if (storage_type_ == "redis") {
     instance = std::make_unique<RedisInternalKV>(GetRedisClientOptions());
   } else if (storage_type_ == "memory") {
-    instance = std::make_unique<MemoryInternalKV>(main_service_);
+    instance = std::make_unique<StoreClientInternalKV>(
+        std::make_unique<InMemoryStoreClient>(main_service_));
   }
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:452]: Why not just use store clients here? Also for redis.

**Review by fishbone — COMMENTED** [review]
Could you remind me why we can't build table on top of internal kv?

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 1 above)
Comment (scv119) [reply src/ray/gcs/gcs_server/gcs_server.cc:452]: yeah I think this will mostly impact Ant. cc @raulchen to see if we should turn it on by default.

**Comment by scv119** [comment]
> Could you remind me why we can't build table on top of internal kv?
Yeah I think it's because of `BatchDelete` and `GetNextJobID`

**Comment by iycheng** [comment]
My feeling is that we should hide the underline store from internal kv and it should be built upon StoreClient directly. 

In the end, the store client will be used by table and kv. We can use memory or Redis as the backend of the store client.

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**